core/src/main/scala/kafka/network/RequestConvertToJson.scala

```diff
@@ -0,0 +1,360 @@
+    node.set("requestHeader", requestHeaderNode(header))
+    node.set("request", request(req, verbose))
+    node
+  }
+
+  /**
+   * Temporary until switch to use the generated schemas.
+   */
+  def offsetsForLeaderEpochRequestNode(request: OffsetsForLeaderEpochRequest, version: Short): JsonNode = {
+    val node = new ObjectNode(JsonNodeFactory.instance)
+    if (version >= 3) {
+      node.set("replicaId", new IntNode(request.replicaId))
+    }
+    val topics = CollectionUtils.groupPartitionDataByTopic(request.epochsByTopicPartition)
+    val topicsArray = new ArrayNode(JsonNodeFactory.instance)
+    topics.forEach { (topicName, partitions) =>
+      val topicNode = new ObjectNode(JsonNodeFactory.instance)
+      topicNode.set("name", new TextNode(topicName))
+      val partitionsArray = new ArrayNode(JsonNodeFactory.instance)
+      partitions.forEach { (partitionIndex, partitionData) =>
+        val partitionNode = new ObjectNode(JsonNodeFactory.instance)
+        partitionNode.set("partitionIndex", new IntNode(partitionIndex))
+        if (version >= 2) {
+          val leaderEpoch = partitionData.currentLeaderEpoch
+          partitionNode.set("currentLeaderEpoch", new IntNode(leaderEpoch.orElse(RecordBatch.NO_PARTITION_LEADER_EPOCH)))
+        }
+        partitionNode.set("leaderEpoch", new IntNode(partitionData.leaderEpoch))
+        partitionsArray.add(partitionNode)
+      }
+      topicNode.set("partitions", partitionsArray)
+      topicsArray.add(topicNode)
+    }
+    node.set("topics", topicsArray)
+    node
+  }
+
+  /**
+   * Temporary until switch to use the generated schemas.
+   */
+  def produceRequestNode(request: ProduceRequest, version: Short): JsonNode = {
+    val node = new ObjectNode(JsonNodeFactory.instance)
+    if (version >= 3) {
+      if (request.transactionalId == null) {
+        node.set("transactionalId", NullNode.instance)
+      } else {
+        node.set("transactionalId", new TextNode(request.transactionalId))
+      }
+    }
+    node.set("acks", new ShortNode(request.acks))
+    node.set("timeoutMs", new IntNode(request.timeout))
+    val topics = CollectionUtils.groupPartitionDataByTopic(request.partitionRecordsOrFail())
+    val topicsArray = new ArrayNode(JsonNodeFactory.instance)
+    topics.forEach { (topicName, partitions) =>
+      val topicNode = new ObjectNode(JsonNodeFactory.instance)
+      topicNode.set("name", new TextNode(topicName))
+      val partitionsArray = new ArrayNode(JsonNodeFactory.instance)
+      partitions.forEach { (partitionIndex, partitionData)=>
+        val partitionNode = new ObjectNode(JsonNodeFactory.instance)
+        partitionNode.set("partitionIndex", new IntNode(partitionIndex))
+        if (partitionData == null)
+          partitionNode.set("records", NullNode.instance)
+        else
+          partitionNode.set("records", new BinaryNode(util.Arrays.copyOf(partitionData.buffer().array(), partitionData.validBytes())))
```

Comment:
The `ProduceRequest` does not match the conditions for the change made in the automated protocol. The `records` field is of type `bytes` in `ProduceRequest`, whereas the change to the protocol only effects fields of type `records` like in `FetchResponse`.

Reply:
This ends up breaking `testFormatOfProduceRequestNode()` because it can't deserialize and serialize the data because of the inconsistency problem discussed below.

Reply:
Right. We need to update that test for the time being until we migrate the produce request/response. We may be able to verify the output differently.